### ISSUE TITLE
[source-mongodb] Incremental sync error on specific collection - Saved offset is not valid

### ISSUE THREAD

**Issue by Hugoch** [airbytehq/airbyte#48496]
### Connector Name

source-mongodb

### Connector Version

1.5.12

### What step the error happened?

During the sync

### Relevant information

After a successful initial sync, incremental replication randomly fails after 1-10 successful ones with a 
`Saved offset is not valid. Please reset the connection, and then increase oplog retention and/or increase sync frequency to prevent this from happening in the future`.

I manually checked the resume token from the logs which is invalid, but manually testing a resume token from a previous successful sync works (which seems to confirm that oplog is not truncated).

Reducing the replication frequency to 5 min (with oplog ~4days retention) does not fix.

Some other collections from the same cluster sync without issues.

**More info**
- MongoDB 7.0.15 (Atlas-managed) / ReplicaSet
- Collection:
```
Storage Size: 
78.14MB
Logical Data Size: 
217.94MB
Total Documents:
39641
Indexes Total Size:
14.07MB
```
- Oplog configuration
![Image](https://github.com/user-attachments/assets/b51e7a3a-a4ad-4a68-89fc-1ea00b044753)
- Logs
[failed_sync.log](https://github.com/user-attachments/files/17746641/failed_sync.log)
[mongo_db.log](https://github.com/user-attachments/files/17746643/mongo_db.log)
[ok_sync.log](https://github.com/user-attachments/files/17746642/ok_sync.log)

cc @theyueli 

### Relevant log output

```shell
2024-11-04 18:30:19 [44msource[0m > INFO main i.a.i.s.m.c.MongoDbDebeziumStateUtil(isValidResumeToken):124 Exception : Command failed with error 280 (ChangeStreamFatalError): 'PlanExecutor error during aggregation :: caused by :: cannot resume stream; the resume token was not found. {_data: "82672911AA000000042B042C0100296E5A1004111CD8A7C3AC4108BD652688BF9B4961463C6F7065726174696F6E54797065003C7570646174650046646F63756D656E744B65790046645F696400646607170ADC704B0B4CEF9E4F000004"}' on server foo-prod-shard-00-02.abcde.mongodb.net:27017. The full response is {"errorLabels": ["NonResumableChangeStreamError"], "ok": 0.0, "errmsg": "PlanExecutor error during aggregation :: caused by :: cannot resume stream; the resume token was not found. {_data: \"82672911AA000000042B042C0100296E5A1004111CD8A7C3AC4108BD652688BF9B4961463C6F7065726174696F6E54797065003C7570646174650046646F63756D656E744B65790046645F696400646607170ADC704B0B4CEF9E4F000004\"}", "code": 280, "codeName": "ChangeStreamFatalError", "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1730745019, "i": 5}}, "signature": {"hash": {"$binary": {"base64": "8gooMnQDcDZ4N47zoh55j9GzBjk=", "subType": "00"}}, "keyId": 7428267533251117062}}, "operationTime": {"$timestamp": {"t": 1730745019, "i": 4}}}
2024-11-04 18:30:19 [44msource[0m > INFO main i.a.i.s.m.c.MongoDbDebeziumStateUtil(isValidResumeToken):125 Invalid resume token '826729119A0000003D2B042C0100296E5A100479290C5D090D4810962649B5F651BDB3463C6F7065726174696F6E54797065003C7570646174650046646F63756D656E744B65790046645F6964006466BDC93807F670F0E483042F000004' present, corresponding to timestamp (seconds after epoch) : 1730744730, due to reason Command failed with error 280 (ChangeStreamFatalError): 'PlanExecutor error during aggregation :: caused by :: cannot resume stream; the resume token was not found. {_data: "82672911AA000000042B042C0100296E5A1004111CD8A7C3AC4108BD652688BF9B4961463C6F7065726174696F6E54797065003C7570646174650046646F63756D656E744B65790046645F696400646607170ADC704B0B4CEF9E4F000004"}' on server foo-prod-shard-00-02.abcde.mongodb.net:27017. The full response is {"errorLabels": ["NonResumableChangeStreamError"], "ok": 0.0, "errmsg": "PlanExecutor error during aggregation :: caused by :: cannot resume stream; the resume token was not found. {_data: \"82672911AA000000042B042C0100296E5A1004111CD8A7C3AC4108BD652688BF9B4961463C6F7065726174696F6E54797065003C7570646174650046646F63756D656E744B65790046645F696400646607170ADC704B0B4CEF9E4F000004\"}", "code": 280, "codeName": "ChangeStreamFatalError", "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1730745019, "i": 5}}, "signature": {"hash": {"$binary": {"base64": "8gooMnQDcDZ4N47zoh55j9GzBjk=", "subType": "00"}}, "keyId": 7428267533251117062}}, "operationTime": {"$timestamp": {"t": 1730745019, "i": 4}}}
2024-11-04 18:30:19 [46mreplication-orchestrator[0m > SOURCE analytics [airbyte/source-mongodb-v2:1.5.12] | Type: db-sources-cdc-cursor-invalid | Value: 1
2024-11-04 18:30:19 [44msource[0m > ERROR main i.a.i.s.m.MongoDbSource(read):157 Unable to perform sync read operation. io.airbyte.commons.exceptions.ConfigErrorException: Saved offset is not valid. Please reset the connection, and then increase oplog retention and/or increase sync frequency to prevent his from happening in the future. See https://docs.airbyte.com/integrations/sources/mongodb-v2#mongodb-oplog-and-change-streams for more details
	at io.airbyte.integrations.source.mongodb.cdc.MongoDbCdcInitializer.createCdcIterators(MongoDbCdcInitializer.java:136) ~[io.airbyte.airbyte-integrations.connectors-source-mongodb-v2.jar:?]
	at io.airbyte.integrations.source.mongodb.MongoDbSource.read(MongoDbSource.java:149) [io.airbyte.airbyte-integrations.connectors-source-mongodb-v2.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.readSerial(IntegrationRunner.kt:290) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.runInternal(IntegrationRunner.kt:190) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run(IntegrationRunner.kt:119) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run$default(IntegrationRunner.kt:113) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run(IntegrationRunner.kt) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.integrations.source.mongodb.MongoDbSource.main(MongoDbSource.java:53) [io.airbyte.airbyte-integrations.connectors-source-mongodb-v2.jar:?]

Stack Trace: io.airbyte.commons.exceptions.ConfigErrorException: Saved offset is not valid. Please reset the connection, and then increase oplog retention and/or increase sync frequency to prevent his from happening in the future. See https://docs.airbyte.com/integrations/sources/mongodb-v2#mongodb-oplog-and-change-streams for more details
	at io.airbyte.integrations.source.mongodb.cdc.MongoDbCdcInitializer.createCdcIterators(MongoDbCdcInitializer.java:136)
	at io.airbyte.integrations.source.mongodb.MongoDbSource.read(MongoDbSource.java:149)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.readSerial(IntegrationRunner.kt:290)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.runInternal(IntegrationRunner.kt:190)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run(IntegrationRunner.kt:119)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run$default(IntegrationRunner.kt:113)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run(IntegrationRunner.kt)
	at io.airbyte.integrations.source.mongodb.MongoDbSource.main(MongoDbSource.java:53)

2024-11-04 18:30:19 [44msource[0m > ERROR main i.a.c.i.u.ConnectorExceptionHandler(handleException):68 caught exception! io.airbyte.commons.exceptions.ConfigErrorException: Saved offset is not valid. Please reset the connection, and then increase oplog retention and/or increase sync frequency to prevent his from happening in the future. See https://docs.airbyte.com/integrations/sources/mongodb-v2#mongodb-oplog-and-change-streams for more details
	at io.airbyte.integrations.source.mongodb.cdc.MongoDbCdcInitializer.createCdcIterators(MongoDbCdcInitializer.java:136) ~[io.airbyte.airbyte-integrations.connectors-source-mongodb-v2.jar:?]
	at io.airbyte.integrations.source.mongodb.MongoDbSource.read(MongoDbSource.java:149) ~[io.airbyte.airbyte-integrations.connectors-source-mongodb-v2.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.readSerial(IntegrationRunner.kt:290) ~[airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.runInternal(IntegrationRunner.kt:190) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run(IntegrationRunner.kt:119) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run$default(IntegrationRunner.kt:113) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run(IntegrationRunner.kt) [airbyte-cdk-core-0.45.1.jar:?]
	at io.airbyte.integrations.source.mongodb.MongoDbSource.main(MongoDbSource.java:53) [io.airbyte.airbyte-integrations.connectors-source-mongodb-v2.jar:?]

Stack Trace: io.airbyte.commons.exceptions.ConfigErrorException: Saved offset is not valid. Please reset the connection, and then increase oplog retention and/or increase sync frequency to prevent his from happening in the future. See https://docs.airbyte.com/integrations/sources/mongodb-v2#mongodb-oplog-and-change-streams for more details
	at io.airbyte.integrations.source.mongodb.cdc.MongoDbCdcInitializer.createCdcIterators(MongoDbCdcInitializer.java:136)
	at io.airbyte.integrations.source.mongodb.MongoDbSource.read(MongoDbSource.java:149)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.readSerial(IntegrationRunner.kt:290)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.runInternal(IntegrationRunner.kt:190)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run(IntegrationRunner.kt:119)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run$default(IntegrationRunner.kt:113)
	at io.airbyte.cdk.integrations.base.IntegrationRunner.run(IntegrationRunner.kt)
	at io.airbyte.integrations.source.mongodb.MongoDbSource.main(MongoDbSource.java:53)

2024-11-04 18:30:19 [46mreplication-orchestrator[0m > readFromSource: source exception
io.airbyte.workers.internal.exception.SourceException: Source process exited with non-zero exit code 1
	at io.airbyte.workers.general.BufferedReplicationWorker.readFromSource(BufferedReplicationWorker.java:364) ~[io.airbyte-airbyte-commons-worker-1.1.0.jar:?]
	at io.airbyte.workers.general.BufferedReplicationWorker.lambda$runAsyncWithHeartbeatCheck$3(BufferedReplicationWorker.java:222) ~[io.airbyte-airbyte-commons-worker-1.1.0.jar:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
```

### Contribute

- [ ] Yes, I want to contribute

**Comment by marcosmarxm** [comment]
Thanks for reporting the issue @Hugoch .

@airbytehq/dbsources can someone take a look into this issue?

**Comment by rodireich** [comment]
Hi @Hugoch 
thanks for reaching out.
Do you see any schema change happening before these failures?

Would you be able to share a successive logs from the an initial sync that finished successfully and an incremental sync the failed afterwards?

**Comment by rodireich** [comment]
The following cloud [workspace](https://cloud.airbyte.com/workspaces/7aafa381-2eba-416e-8135-ddc383b7c2cd/connections/ba6a31d1-894c-48c6-b0a3-85c691ad2c89/timeline) may be showing the same error

**Comment by theyueli** [comment]
hi @rodireich, Hugo shared the logs through our email conversations where you have been copied.

I'm attaching the logs here. 

[ok_sync.log](https://github.com/user-attachments/files/17754277/ok_sync.log)
[failed_sync (1).log](https://github.com/user-attachments/files/17754281/failed_sync.1.log)
[mongo_db.log](https://github.com/user-attachments/files/17754283/mongo_db.log)

**Comment by theyueli** [comment]
@rodireich @xiaohansong  
This issue could be related:
https://github.com/airbytehq/airbyte/issues/48435

**Comment by ngocquang2797** [comment]
This is a [similar issue](https://github.com/airbytehq/airbyte/issues/46569), but it hasn’t been resolved yet.
I’m also encountering a similar issue with the error “saved offset is not valid” when syncing data every 30 minutes. I checked the replication info and found that the log length from start to end is approximately 72 hours.
![Image](https://github.com/user-attachments/assets/7755c819-2410-4747-95f0-3c27322a217f)

![Image](https://github.com/user-attachments/assets/5482dda0-7b2f-429b-a909-ee8846090118)
I increased the sync frequency to every 6 minutes, but I’m still encountering this error.

**Comment by ngocquang2797** [comment]
I’m also encountering a similar issue with the error “saved offset is not valid” when syncing data every 30 minutes. I checked the replication info and found that the log length from start to end is approximately 72 hours.
![Image](https://github.com/user-attachments/assets/7755c819-2410-4747-95f0-3c27322a217f)

![Image](https://github.com/user-attachments/assets/5482dda0-7b2f-429b-a909-ee8846090118)
I increased the sync frequency to every 6 minutes, but I’m still encountering this error.

**Comment by Hugoch** [comment]
@ngocquang2797 It seems that your actual Oplog size is bigger than your configured Oplog size. This may be due to your majority commit point being far in the past, so you probably have a large replication lag? I am not sure it is related to the issue, but it might be worth investigating.

**Comment by theyueli** [comment]
@Hugoch @ngocquang2797 we created a test image that could potentially fix this bug (in the PR mentioned above), and we created a dev image for you to try, could you set your MongoDB source connector to version 1.5.12-dev.2ba32d767b 

please reset the connection, and test again, hopefully this resolves the issue. Thanks!

**Comment by Hugoch** [comment]
Hi @theyueli , thanks for the patch! I tried with the new connector without success. I tried with and without schema enforced.

![Image](https://github.com/user-attachments/assets/1753b62e-dc44-4c4a-97c5-254481ea8794)

Here are the failing logs
[fail.log](https://github.com/user-attachments/files/17948798/fail.log)

**Comment by theyueli** [comment]
@postamar could you take another look at this issue?

**Comment by aya-z-arketa** [comment]
hey! @theyueli is there any update on this issue? Experiencing the same issue.

**Comment by aya-z-arketa** [comment]
@rodireich @yardencarmeli any update on this issue? experiencing it as well. thanks!

**Comment by haithem-souala** [comment]
Any update?

**Comment by phammer** [comment]
Hi all, 

sharing some learnings. We were suffering of this problem as well for quite a long time. And recently, found a little warning in the logs that hinted to the actual problem: the connector cannot read the oplog size of our MongoDB, and simply falls back to some default (we don't know what that is). 

In our case, the DB and thus OpLog is big, and w/o the connector being able to see that, we always ended up with the problem mention in the issue description.

**Log Line**

This log warning gave the clue. 

```
2025-05-04 01:00:16 source WARN main i.a.i.s.m.c.MongoDbCdcInitializer(logOplogInfo):280 Unable to query for op log stats, exception: {}Command failed with error 13 (Unauthorized): ‘not authorized on local to execute command { collStats: “[oplog.rs](http://oplog.rs/)”, $db: “local”, lsid: { id: UUID(“a5026a57-7496-48ee-b752-a51f027ba4c7”) } }’ on server trip-builder-prod-shard-00-02.uy4pw.mongodb.net:27017. The full response is {“ok”: 0.0, “errmsg”: “not authorized on local to execute command { collStats: \“[oplog.rs](http://oplog.rs/)\“, $db: \“local\“, lsid: { id: UUID(\“a5026a57-7496-48ee-b752-a51f027ba4c7\“) } }“, “code”: 13, “codeName”: “Unauthorized”, “$clusterTime”: {“clusterTime”: {“$timestamp”: {“t”: 1746313216, “i”: 4}}, “signature”: {“hash”: {“$binary”: {“base64": “qBjTO6eIs+yCMLI623lz91tMrG8=“, “subType”: “00"}}, “keyId”: 7445243400078491836}}, “operationTime”: {“$timestamp”: {“t”: 1746313216, “i”: 4}}}
85
```

**Mitigating the Issue**

We gave "read@local" to the Airbyte MongoDB user. Once that is done, the above warning is gone, and instead you'll see something like this:

```
2025-05-05 12:30:16 source INFO main i.a.i.s.m.c.MongoDbCdcInitializer(logOplogInfo):276 Max oplog size is 78383153152 bytes
2025-05-05 12:30:16 source INFO main i.a.i.s.m.c.MongoDbCdcInitializer(logOplogInfo):277 Free space in oplog is 10821718016 bytes
```

The logs now show that the connector was able to retrieve the oplog size and thus, can do its optimization. By doing so, the error mention in the issue description disappeared. 


**Thoughts**

-  The log line should be an ERROR not WARNING. The MongoDB connector doesn't work correctly with an assumed oplog size. It needs the right one.

- Use Change Streams instead of OpLog? I am not very familiar with the details on the topic. But as far as I've seen, by migrating to the newer change streams features, the permission read@local shouldn't be needed anymore.

PS: In the past, we also had issues with the MongoDB connector. We run our Airbyte on a EC2 instance and part of the solution back then was also to give it more power. Airbyte is not that explicit in telling you when it was a resource issue.

**Comment by ivan-sukhomlyn** [comment]
I have the same issue with the MongoDB source.

Details - https://github.com/airbytehq/airbyte/issues/46569#issuecomment-3072799043

@martasakal1 we also use the `secondaryPreffered` read preference.
Have you been able to fix the issue by changing the preference to primary first?

**Comment by sauravkr20** [comment]
hiii, 

for our case these are the configurations i have set : 
- oplog retention period is 24hrs 
- storage autoscaling is done 
- connection sync is running every 2 hours 

airbyte 18 version

but still the same issue.... when are we expecting it to be fixed

**Comment by ivan-sukhomlyn** [comment]
@sauravkr20 There's a potential fix at the `2.0.4-dev.5b6b076be4` version of the MongoDB source - https://github.com/airbytehq/airbyte/issues/46569#issuecomment-3259404963

**Comment by yardencarmeli** [comment]
Hi all! The v2.0.4 release is out and should now be available. Since this ticket was opened almost a year ago, it likely doesn’t reflect the current version anymore (we’ve had over 15 releases since then).

I’ll keep this ticket open until 10/1 for anyone who wants to upgrade to v2.0.4 and share feedback. If you run into issues that are not related to the “2 states” bug fixed in this version, please open a separate ticket so we can review them properly.

**Comment by raphaelsampaio** [comment]
2.0.4 is working fine for us
